rust/native/src/cvt.rs

```diff
@@ -0,0 +1,37 @@
+use std::io;
+
+#[doc(hidden)]
+pub trait IsMinusOne {
+    fn is_minus_one(&self) -> bool;
+}
+
+macro_rules! impl_is_minus_one {
+    ($($t:ident)*) => ($(impl IsMinusOne for $t {
+        fn is_minus_one(&self) -> bool {
+            *self == -1
+        }
+    })*)
+}
+
+impl_is_minus_one! { i8 i16 i32 i64 isize }
+
+pub(crate) fn cvt<T: IsMinusOne>(t: T) -> io::Result<T> {
+    if t.is_minus_one() {
+        Err(io::Error::last_os_error())
+    } else {
+        Ok(t)
+    }
+}
+
+pub(crate) fn cvt_r<T, F>(mut f: F) -> io::Result<T>
+where
+    T: IsMinusOne,
+    F: FnMut() -> T,
+{
+    loop {
+        match cvt(f()) {
+            Err(ref e) if e.kind() == io::ErrorKind::Interrupted => {}
+            other => return other,
+        }
+    }
+}
```

Comment:
I'm really confused about this entire file :D

What does `cvt` stand for? If I understand correctly (and to be honest I haven't yet spent too much time on it) it's basically used to see whether a function called has returned -1 to signal an error? Do we really need this level of abstraction? I'm probably missing something!

Reply:
cvt* macro is used all over the place in rust system code to handle and return system errors. I think *t!() macros "stands" for test in general. and "_r" for retry/repeat. So "cv" is perhaps "call value"? Yes, it is really nice to have. You'll be able to judge in the diff.

Reply:
@nertpinx digged up [this nice article](https://people.gnome.org/~federico/blog/index7.html) which explains things in a lot of detail, including the need for `cvt`.

Anyway, if we're copying stuff from Rust's `std` we need at the very least to give proper attribution and maintain the copyright clause; the licenses are compatible so no problem there.